scripts/lib/LinkChecker.ts

```diff
@@ -0,0 +1,128 @@
+// obtain a copy of this license in the LICENSE file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+import path from "node:path";
+
+const DOCS_ROOT = "./docs";
+const CONTENT_FILE_EXTENSIONS = [".md", ".mdx", ".ipynb"];
+
+export class File {
+  readonly path: string;
+  anchors: string[];
+
+  constructor(path: string) {
+    this.path = path;
+    this.anchors = [];
+  }
+
+  setAnchors(anchors: string[]) {
+    this.anchors = anchors;
+  }
+}
+
+export class Link {
+  readonly value: string;
+  readonly anchor: string;
+  readonly origins: string[];
+  readonly isExternal: boolean;
+
+  constructor(linkString: string, origins: string[]) {
+    /*
+     * linkString: Link as it appears in source file
+     *    origins: Paths to source file containing link
+     */
+
+    const splitLink = linkString.split("#", 2);
+    this.value = splitLink[0];
+    this.anchor = splitLink.length > 1 ? `#${splitLink[1]}` : "";
+    this.origins = origins;
+    this.isExternal = linkString.startsWith("http");
+  }
+
+  resolve(origin: string): string[] {
+    /*
+     * Return list of possible paths link could resolve to
+     */
+    if (this.isExternal) {
+      return [this.value];
+    }
+    if (this.value === "") {
+      return [origin];
+    } // link is just anchor
+    if (this.value.startsWith("/images")) {
+      return [path.join("public/", this.value)];
+    }
+
+    let baseFilePath;
+    if (this.value.startsWith("/")) {
+      // Path is relative to DOCS_ROOT
+      baseFilePath = path.join(DOCS_ROOT, this.value);
+    } else {
+      // Path is relative to origin file
+      baseFilePath = path.join(path.dirname(origin), this.value);
+    }
+    // Remove trailing '/' from path.join
+    baseFilePath = baseFilePath.replace(/\/$/gm, "");
+
+    // File may have one of many extensions (.md, .ipynb etc.), and/or be
+    // directory with an index file (e.g. `docs/build` should resolve to
+    // `docs/build/index.mdx`). We return a list of possible filenames.
+    let possibleFilePaths = [];
+    for (let index of ["", "/index"]) {
+      for (let extension of CONTENT_FILE_EXTENSIONS) {
+        possibleFilePaths.push(baseFilePath + index + extension);
+      }
+    }
+    return possibleFilePaths;
+  }
+
+  checkExternalLink() {
+    // External link checking not supported yet
+    return true;
+  }
+
+  checkInternalLink(existingFiles: File[], origin: string) {
+    /*
+     * True if link is in `existingFiles`, otherwise false
+     */
+    const possiblePaths = this.resolve(origin);
+    for (let filePath of possiblePaths) {
+      if (existingFiles.some((file) => file.path == filePath)) {
+        return true;
+      }
+    }
+    return false;
```

Comment:
```suggestion
    return possiblePaths.some((possibleFile) => (
       existingFiles.some((exitingFile) => existingFile.path == possibleFile)
      )
    )
```

Reply:
You could also get the set intersection. If there are any elements in both sets, return true. That's probably easier to read and should be more efficient.